srk/src/almostCommuting.ml

```diff
@@ -0,0 +1,146 @@
+open Linear
+
+module VS = QQVectorSpace
+
+type kind = Commute | Reset | Ignore
+
+type phased_segment = {
+  sim1 : QQMatrix.t;
+  sim2 : QQMatrix.t;
+  phase1 : QQMatrix.t array;
+  phase2 : (kind * QQMatrix.t) array
+}
+
+type phased_segmentation = phased_segment list
+
+let commuting_space mA mB =
+  let dims = SrkUtil.Int.Set.elements (QQMatrix.row_set mA) in
+  let mAB = QQMatrix.mul mA mB in
+  let mBA = QQMatrix.mul mB mA in
+  let mC = QQMatrix.add mAB (QQMatrix.scalar_mul (QQ.negate QQ.one) mBA) in
+  nullspace (QQMatrix.transpose mC) dims
+
+let intersect_rowspaces matrices dims =
+  Array.fold_left 
+    (fun mA mB -> 
+      let (mC, _) = intersect_rowspace mA mB in
+      QQMatrix.mul mC mA)
+    (QQMatrix.identity dims)
+    matrices
+
+let rowspace_equal mA mB =
+  VS.equal (VS.of_matrix mA) (VS.of_matrix mB)
+
+let commuting_segment matrices dims =
+  let pairs = BatArray.cartesian_product matrices matrices in
+  let cspaces = Array.map (fun (mA, mB) -> VS.matrix_of (commuting_space mA mB)) pairs in
+  let mS = intersect_rowspaces cspaces dims in
+  let rec fix mS =
+    let maxlds = Array.map (fun mat -> max_lds mS (QQMatrix.mul mS mat)) matrices in
+    let sims, matr = BatArray.split maxlds in
+    let mSS = intersect_rowspaces 
+                (Array.map (fun m -> QQMatrix.mul m mS) sims)
+                dims
+    in
+    if rowspace_equal mS mSS then
+      mS, matr
+    else
+      fix mSS
+  in
+  fix mS
+
+let iter_all     = Array.map (fun (_, m) -> m)
+let iter_reset   = BatArray.filter_map (fun (k, m) -> if k == Reset then Some m else None)
+let iter_commute = BatArray.filter_map (fun (k, m) -> if k == Commute then Some m else None)
+
+module PhasedSegment = struct
+
+  type t = phased_segment
+
+  (* TODO: find proper representation *)
+  let show p = QQMatrix.show p.sim1
+
+  let equal p q =
+    QQMatrix.equal p.sim1 q.sim1 &&
+    QQMatrix.equal p.sim2 q.sim2 &&
+    BatArray.for_all2
+      QQMatrix.equal
+      p.phase1
+      q.phase1 &&
+    BatArray.for_all2
+      (fun (k1, m1) (k2, m2) -> k1 == k2 && QQMatrix.equal m1 m2)
+      p.phase2
+      q.phase2
+
+  let make pairs =
+    if Array.length pairs == 0 then
+      raise (Invalid_argument "Array of matrices should not be empty")
+    else
+      let _, mA = Array.get pairs 0 in
+      let dims = SrkUtil.Int.Set.elements (QQMatrix.column_set mA) in
+      let mS, phase1 = commuting_segment (iter_all pairs) dims in
+      let mT, _ = commuting_segment (iter_commute pairs) dims in
+      let rec fix mT =
+        let ls_maxlds = Array.map
+          (fun (k, m) ->
+            if k == Reset then
+              max_lds ~zero_rows:true mS (QQMatrix.mul mT m)
```

Comment:
This doesn't look right to me -- in general, if k = reset, then we should expect the associated matrix to be m x n, where m is the dimension of the almost commuting space and n is the dimension of the commuting space (max_lds will always produce a square matrix + a simulation).

Reply:
Assuming we call `max_lds mS mC` where S is a n x m matrix and C an o x m matrix with n <= o <=m. Then this gives me a simulation T and a matrix C' s.t. TC = C'TS.
If I understood the code of `max_lds` correctly, then `max_lds` implicitly treats its input matrices as square matrices, and then removes coordinates which correspond to zero rows in T*S. 
For instance the following
```ocaml
let mS = mk_matrix [[0; 1; 0]; 
                    [0; 0; 1]] 
in
let mC = mk_matrix [[0; 0; 2];
                    [0; 1; 2];
                    [0; 0; 1]]
in
max_lds mS mC
```
produces a simulation `[[1; 0]; [0; 1]]` and a square matrix `[[0; 2]; [1; 2]]`.
However, if we do not remove the coordinates corresponding to zero rows in TS (i.e. `max_lds ~zero_rows:true mS mC`), we get a simulation `[[1; 0; 0]; [0; 1; 0]; [0; 0; 1]` and a matrix C' (with implicit zeroes in the third column)
```ocaml
[[0; 2]; 
 [1; 2]; 
 [0; 1]]
```
We then have C'S = C.


Another example:
```ocaml
let mS = mk_matrix [[0; 0; 1]] in
let mC = mk_matrix [[0; 0; 2];
                    [0; 0; 1]]
in
max_lds ~zero_rows:true mS mC
```
gives a simulation `[1; 0]; [0; 1]` and a matrix `[[2]; [1]]`. `~zero_rows:false` would give a simulation `[1]` and a matrix `[2]`.

Reply:
In
```ocaml
let mS = mk_matrix [[0; 1; 0]; 
                    [0; 0; 1]] 
in
let mC = mk_matrix [[0; 0; 2];
                    [0; 1; 2];
                    [0; 0; 1]]
in
max_lds mS mC
```
we have a transition system Sx' = Cx, corresponding to
```
y' = 2z
z' = y + 2z
0 = z
```
That is, the transition system is partial (defined only on the subspace where z = 0).  You shouldn't encounter such partial transition systems, because you are given as input a total transition system.
Ge

Reply:
In the case where `k==Reset` I'm not considering a transition system. Instead mS is the simulation where all matrices commute. and mC corresponds to T*A where T is the simulation where the nonreset matrices commute and A is a matrix in the reset part. I'm using `max_lds` to find a T' and a A' s.t. T'TA = A'T'S. If T' is not the idendity, then I have to refine T with T'T, otherwise A'T' would be \rho(A) as in the definition of phased segments.

Reply:
I see... I think there's no need for a fixpoint computation here: you can find the determined functions of the Resets, and then find the commuting space of the Commutes, and then you're done.

Reply:
Right. I forgot to add a comment saying that the fixpoint should be reached after at most two iterations. Anyway, I changed the code to get rid of the unnecessary recursion.

Reply:
Using `max_lds` here, does not work in general. I have to use `pushout`. Already changed.